extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java

```diff
@@ -96,34 +132,43 @@ private void renewLoop() {
     }
     Duration retryPeriod = config.getRetryPeriod();
     long retryPeriodMillis = retryPeriod.toMillis();
-    AtomicBoolean renewed = new AtomicBoolean(true);
-
-    ScheduledFuture scheduledFuture =
-        scheduledWorkers.scheduleAtFixedRate(
-            () -> {
-              Future<Boolean> future = leaseWorkers.submit(this::tryAcquireOrRenew);
-              try {
-                //  set renew success
-                renewed.set(future.get(retryPeriodMillis, TimeUnit.MILLISECONDS));
-              } catch (Throwable t) {
-                future.cancel(true);
-              }
-            },
-            0,
-            retryPeriodMillis,
-            TimeUnit.MILLISECONDS);
+    Duration renewDeadline = config.getRenewDeadline();
+    long renewDeadlineMillis = renewDeadline.toMillis();
 
     try {
-      while (renewed.get()) {
-        if (log.isDebugEnabled()) {
-          log.debug("Successfully renewed lease");
+      while (true) {
+        final Future<Boolean> future =
+            leaseWorkers.submit(
+                () -> {
+                  try {
+                    // retry until success or interrupted
+                    while (!tryAcquireOrRenew()) {
+                      Thread.sleep(retryPeriodMillis);
+                    }
+                  } catch (InterruptedException e) {
+                    return false;
+                  }
+                  return true;
+                });
+
+        boolean renewResult;
+        try {
+          renewResult = future.get(renewDeadlineMillis, TimeUnit.MILLISECONDS);
```

Comment:
applying timeout here is a good way to mock out the channel-based timer's behavior in golang i agree.